url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/geom/LA.java

```diff
@@ -0,0 +1,1062 @@
+			} catch (ArrayIndexOutOfBoundsException ex) {
+				My[c][mCols] = 0d;
+			}
+		}
+
+		for (int r = 0; r < mRows; ++r) {
+			for (int c = 0; c < mCols; ++c) {
+				try {
+					My[r][c] = M[r][c];
+				} catch (ArrayIndexOutOfBoundsException ex) {
+					My[r][c] = 0d;
+				}
+			}
+		}
+
+		try {
+			return solveInplace(My);
+			//						
+			// }catch(SingularException sex){
+			// return rationalSolve(My);
+		} finally {
+		}
+	}
+
+	public static double[] laPacksolve(double[][] M, double[] y)
+			throws SingularException {
+		// double[][] vresult = new double[1][y.length];
+		// for(int i = 0 ; i< y.length ; ++ i ){
+		// vresult[0][i]=y[1];
+		// }
+		double[][] vresult = new double[][] { y.clone() };
+		double[][] vresultt = LA.transpose(vresult);
+		int[] iPiv = new int[y.length];
+		intW info = new intW(1);
+		DGESV.DGESV(/* int n */M.length,
+		/* int nrhs */1,
+		/* double[][]a */M,
+		/* int[] ipiv */iPiv,
+		/* double[][] b */vresultt,
+		/* intW info */info);
+
+		if (info.val > 0) {
+			throw new SingularException();
+		}
+
+		double[] result = new double[y.length];
+		for (int i = 0; i < result.length; ++i) {
+			result[i] = vresultt[i][0];
+		}
+
+		return result;
+	}
+
+	public static double[] laPacksolveSym(double[][] M, double[] y)
+			throws SingularException {
+		// double[][] vresult = new double[1][y.length];
+		// for(int i = 0 ; i< y.length ; ++ i ){
+		// vresult[0][i]=y[1];
+		// }
+
+		int N = M.length;
+		int i = 0;
+		double[][] vresult = new double[][] { y.clone() };
+		double[][] vresultt = LA.transpose(vresult);
+		int[] iPiv = new int[y.length];
+		intW info = new intW(1);
+		double[] work = new double[1];
+		int lWork = -1;
+
+		DSYSV.DSYSV(/* String uplo */"U",
+		/* int n */N,
+		/* int nrhs */1,
+		/* double[] ap */M,
+		/* int[] iPiv */iPiv,
+		/* double[][] b */vresultt, work, lWork,
+		/* intW info */info);
+
+		if (info.val > 0) {
+			throw new SingularException();
+		}
+
+		lWork = (int) (work[0]);
+		work = new double[lWork];
+
+		DSYSV.DSYSV(/* String uplo */"U",
+		/* int n */N,
+		/* int nrhs */1,
+		/* double[] ap */M,
+		/* int[] iPiv */iPiv,
+		/* double[][] b */vresultt, work, lWork,
+		/* intW info */info);
+
+		if (info.val > 0) {
+			throw new SingularException();
+		}
+
+		double[] result = new double[y.length];
+		for (i = 0; i < result.length; ++i) {
+			result[i] = vresultt[i][0];
+		}
+
+		return result;
+	}
+
+	public static double[] solveInplace(double[][] My) throws SingularException {
+		return solveInplace(My, false);
+	}
+
+	/**
+	 * A simple in-place Gausssian solver which fails if the matrix of the
+	 * equations system is singular.
+	 * 
+	 * @param My
+	 *            the matrix and the inhomogenious part of the system
+	 * @return an array x of doubles such that Mx = y;
+	 * @throws Exception
+	 *             if the equation system doesn't have exactly one such
+	 *             solution.
+	 */
+	public static double[] solveInplace(double[][] My, boolean safe)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 499-522 into a separate method.